File: SortVision/public/code/merge/lua/mergeSort.lua
-- Merge Sort
-- Time: Best/Avg/Worst O(n log n), Space: O(n)
local function merge(left, right)
  local result = {}
  local i, j = 1, 1

  while i <= #left and j <= #right do
    if left[i] <= right[j] then
      result[#result + 1] = left[i]
      i = i + 1
    else
      result[#result + 1] = right[j]
      j = j + 1
    end
  end

  while i <= #left do result[#result + 1] = left[i]; i = i + 1 end
  while j <= #right do result[#result + 1] = right[j]; j = j + 1 end
  return result
end

local function mergeSort(arr)
  if #arr <= 1 then return arr end

  local mid = math.floor(#arr / 2)
  local left, right = {}, {}
  for i = 1, mid do left[#left + 1] = arr[i] end
  for i = mid + 1, #arr do right[#right + 1] = arr[i] end

  return merge(mergeSort(left), mergeSort(right))
end

local arr = {64, 25, 12, 22, 11}
print(table.concat(mergeSort(arr), ", "))

File: SortVision/public/code/merge/ruby/mergeSort.rb
# Merge Sort
# Time: Best/Avg/Worst O(n log n), Space: O(n)
def merge_sort(arr)
  return arr if arr.length <= 1

  mid = arr.length / 2
  left = merge_sort(arr[0...mid])
  right = merge_sort(arr[mid..])
  merge(left, right)
end

def merge(left, right)
  result = []
  i = j = 0

  while i < left.length && j < right.length
    if left[i] <= right[j]
      result << left[i]
      i += 1
    else
      result << right[j]
      j += 1
    end
  end

  result.concat(left[i..] || []).concat(right[j..] || [])
end

arr = [64, 25, 12, 22, 11]
p merge_sort(arr)
